filament/backend/src/vulkan/memory/Resource.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_BACKEND_VULKAN_MEMORY_RESOURCE_H
+#define TNT_FILAMENT_BACKEND_VULKAN_MEMORY_RESOURCE_H
+
+#include <private/backend/HandleAllocator.h>
+#include <utils/Mutex.h>
+
+#include <type_traits>
+#include <cstdint>
+
+namespace filament::backend::fvkmemory {
+
+using CounterIndex = int32_t;
+using HandleId = HandleBase::HandleId;
+
+class ResourceManager;
+
+template <typename D>
+struct resource_ptr;
+
+// Subclasses of VulkanResource must provide this enum in their construction.
+enum class ResourceType : uint8_t {
+    BUFFER_OBJECT = 0,
+    INDEX_BUFFER = 1,
+    PROGRAM = 2,
+    RENDER_TARGET = 3,
+    SWAP_CHAIN = 4,
+    RENDER_PRIMITIVE = 5,
+    TEXTURE = 6,
+    TEXTURE_STATE = 7,
+    TIMER_QUERY = 8,
+    VERTEX_BUFFER = 9,
+    VERTEX_BUFFER_INFO = 10,
+    DESCRIPTOR_SET_LAYOUT = 11,
+    DESCRIPTOR_SET = 12,
+    FENCE = 13,
+    UNDEFINED_TYPE = 14,    // Must be the last enum because we use it for iterating over the enums.
+};
+
+template<typename D>
+ResourceType getTypeEnum() noexcept;
+
+std::string getTypeStr(ResourceType type);
+
+inline bool isThreadSafeType(ResourceType type) {
+    return type == ResourceType::FENCE || type == ResourceType::TIMER_QUERY;
+}
+
+template<bool THREAD_SAFE = false>
+struct ResourceImpl {
```

Comment:
is it all backend handles that now will inherit from `ResourceImpl` or only those that need reference counting? (Asking because the size overhead is not trivial).

Reply:
yes, for simplicity all backend handles will inherit from `ResourceImpl`. With the previous change where I moved bits from VulkanRenderTarget (the largest handle) to the heap, we might be ok with this increase.

Reply:
but that means that all handles will increase by at least 16 bytes. I think 8 bytes is the minimum (the cound + some meta data). What is really sad is to have to store a pointer. Would it be too painful to pass it around everywhere it's needed? Or maybe we can think of doing that in another change?

Reply:
Yeah, I think for the pointer to the manager, we can definitely revisit.